src/bse_iterative.F

```diff
@@ -0,0 +1,1788 @@
+
+         !Copy eigenvalues for threshold
+         Subspace_prev_eigenval(:) = Subspace_full_eigenval(:num_exc_en)
+
+         DEALLOCATE (AZ, & !BZ,
+                     Z_vectors, M_ia_tmp, M_ji_tmp, RI_vector, &
+                     Subspace_new_eigenval, Subspace_new_eigenvec, Subspace_residuals_reshaped, &
+                     Subspace_add_dir, AZ_reshaped, Z_vectors_reshaped, Subspace_ritzvec, Subspace_full_eigenval, &
+                     Res_norms, En_diffs)
+
+         IF (bse_full_diag_debug) THEN
+            DEALLOCATE (En_diffs_exact)
+         END IF
+
+         !Orthonorm:
+         CALL orthonormalize_W(W_vectors, num_Z_vectors, homo, virtual)
+
+         ! print *,"W_vectors"
+         ! DO k_print=1,num_Z_vectors
+         !    print *,k_print
+         !    print *,W_vectors(:,:,k_print)
+         ! END DO
+
+      END DO
+
+      CALL timestop(handle)
+
+   END SUBROUTINE
+
+! **************************************************************************************************
+!> \brief ...
+!> \param W_vectors ...
+!> \param Z_vectors ...
+!> \param Subspace_add_dir ...
+!> \param Subspace_ritzvec ...
+!> \param num_Z_vectors ...
+!> \param num_new_t ...
+!> \param num_max_z_space ...
+!> \param homo ...
+!> \param virtual ...
+!> \param i_iter ...
+!> \param unit_nr ...
+!> \param res_filename ...
+!> \param en_filename ...
+! **************************************************************************************************
+   SUBROUTINE check_Z_space_dimension(W_vectors, Z_vectors, Subspace_add_dir, Subspace_ritzvec, &
+                                      num_Z_vectors, num_new_t, num_max_z_space, homo, virtual, i_iter, unit_nr, &
+                                      res_filename, en_filename)
+
+      REAL(KIND=dp), ALLOCATABLE, DIMENSION(:, :, :)     :: W_vectors, Z_vectors, Subspace_add_dir, &
+                                                            Subspace_ritzvec
+      INTEGER                                            :: num_Z_vectors, num_new_t, &
+                                                            num_max_z_space, homo, virtual, &
+                                                            i_iter, unit_nr
+      CHARACTER(LEN=100)                                 :: res_filename, en_filename
+
+      CHARACTER(LEN=*), PARAMETER :: routineN = 'check_Z_space_dimension'
+
+      INTEGER                                            :: handle
+
+      CALL timeset(routineN, handle)
+
+      IF (num_Z_vectors + num_new_t .LE. num_max_z_space) THEN
+         W_vectors(:, :, :num_Z_vectors) = Z_vectors(:, :, :)
+         W_vectors(:, :, num_Z_vectors + 1:) = Subspace_add_dir
+         num_Z_vectors = num_Z_vectors + num_new_t
+      ELSE
+         IF (unit_nr > 0) THEN
+            WRITE (unit_nr, *) "MG: Resetting dimension in i_iter=", i_iter
+         END IF
+         IF (unit_nr > 0) THEN
+            OPEN (10, file=res_filename, status="old", access="sequential", position="append")
```

Comment:
Please CP2K for a unit number (close_file/open_file/get_unit_number from the module "cp_files") or use the provide ones (default_input_unit/default_output_unit from the module "machine")